packages/server/src/games.ts

```diff
@@ -58,3 +58,99 @@ export function updateGameState(io: IOServer, store: Store, roomName: string, ga
     io.in(roomName).emit(GameEvent.CurrentState, updatedGameState);
   }
 }
+
+/**
+ * Calculate the immediate new game state, as a result of a player's action. This is not necessarily the final state
+ * but an intermediate one to help the following calculations.
+ */
+function calculatePlayerAction(oldState: GameState, playerAction: PlayerAction) {
+  const activePlayer = oldState.players.filter((playerState) => playerState.username === playerAction.playerName)[0];
+  // TODO: Later on we will need to add order of players, built into the model
+  const opponent = oldState.players.filter((playerState) => playerState.username !== playerAction.playerName)[0];
+  const hand = activePlayer.hand.filter((c) => cardKey(c) !== playerAction.card);
+  const table = [...oldState.table, fromCardKey(playerAction.card)];
```

Comment:
I'm not sure about server side code, but I believe we can mutate the things directly here instead of recreating arrays and objects over and over. Not sure about best practices, what do you think?

So in case this will be `oldState.table.push(oldState)` or well, in this case we will just have 1 state and we keep pushing mutating it until we will return but I'm not sure what's the best practice here, in the client I would never do that

Reply:
Yeah, I am fine with that... We're not talking about huge structures here, but no reason not to do what you say. Will update. As I say in the comment below, I do prefer starting from a clean state. So have a look at this now.

(I guess you also meant `.push(fromCardKey(playerAction.card))`